reference/decoder.go

```diff
@@ -0,0 +1,287 @@
+//
+// Copyright 2019 Insolar Technologies GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package reference
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type ByteDecodeFunc func(s string, target io.ByteWriter) (stringRead int, err error)
+type IdentityDecoder func(base *Global, name string) *Global
+
+type DecoderOptions uint8
+
+const (
+	AllowLegacy DecoderOptions = 1 << iota
+	AllowRecords
+	IgnoreParity
+)
+
+type GlobalDecoder interface {
+	Decode(ref string) (Global, error)
+}
+
+type decoder struct {
+	byteDecoderFactory ByteDecoderFactory
+	legacyDecoder      ByteDecodeFunc
+	defaultDecoder     ByteDecodeFunc
+
+	nameDecoder IdentityDecoder
+	options     DecoderOptions
+}
+
+func NewDecoder(options DecoderOptions) GlobalDecoder {
+	byteDecoderFactory := NewByteDecoderFactory()
```

Comment:
1. Factory must be a parameter pls
2. For convenience - create NewDefaultDecoder, that provide the default factory

Reply:
fixed